### PR TITLE
Cluster Generator reconciling on cluster secret events

### PR DESCRIPTION
**Pull request by dgoodwin**
This PR enables the cluster generator to run and locate all cluster secrets with matching labels. The controller now watches for changes to cluster Secrets and reconciles all appsets with a Cluster generator.

### CODE REVIEW

File: examples/clusters.yaml [hunk 1]
```diff
@@ -10,13 +10,17 @@ metadata:
   name: guestbook
 spec:
   generators:
-  - clusters: {}
+  - clusters:
+      selector:
+        matchLabels:
+          argocd.argoproj.io/secret-type: cluster
```
Comment (dgoodwin) [inline examples/clusters.yaml:16]: This example just matches *every* cluster argo knows about as we're re-using the argo label.

File: pkg/controllers/applicationset_controller.go [hunk 2]
```diff
@@ -49,14 +54,24 @@ func (r *ApplicationSetReconciler) Reconcile(req ctrl.Request) (ctrl.Result, err
 		return ctrl.Result{}, client.IgnoreNotFound(err)
 	}
 
-	var generator generators.Generator
-	generator = generators.NewListGenerator()
+	listGenerator := generators.NewListGenerator()
+	clusterGenerator := generators.NewClusterGenerator(r.Client)
```
Comment (dgoodwin) [inline pkg/controllers/applicationset_controller.go:58]: This portion will conflict with #7 but I can fix up once it merges.

File: pkg/controllers/applicationset_controller.go [hunk 3]
```diff
@@ -49,14 +54,24 @@ func (r *ApplicationSetReconciler) Reconcile(req ctrl.Request) (ctrl.Result, err
...
+
+	// desiredApplications is the main list of all expected Applications from all generators in this appset.
+	var desiredApplications  []argov1alpha1.Application
 	for _, tmpGenerator := range applicationSetInfo.Spec.Generators {
-		desiredApplications, err := generator.GenerateApplications(&tmpGenerator, &applicationSetInfo)
-		log.Infof("desiredApplications %+v", desiredApplications)
+		var apps []argov1alpha1.Application
```
Comment (xianlubird) [inline pkg/controllers/applicationset_controller.go:63]: If we have many kinds of generator,  will this increase many `if else` struct .

should we provide a abstract function to hide these like factory.

Comment (dgoodwin) [reply pkg/controllers/applicationset_controller.go:63]: Per slack it sounds like perhaps we should instantiate all generators we know of, then call each with the AppSet, and let it return us the list of Applications for every item in the list that matches.

Although we should probably discuss if it makes sense to have multiple generators in one AppSet at some point, I'm a little worried that'll never be used due to the nature of the templating and the available params being different.

Should we defer this and followup PR once we know what to do?

Comment (xianlubird) [reply pkg/controllers/applicationset_controller.go:63]: yes.  fine

**Pull request merged**